cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -361,7 +395,36 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
         return Optional.of(provisioningService.deprovision(programRunId, datasetContext));
       case DEPROVISIONED:
-        appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        RunRecordMeta meta = appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        if (meta == null || meta.getCluster() == null) {
+          // this should not happen since when the cluster info should be there after deprovision
+          return Optional.empty();
+        }
+        Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                           meta.getSystemArgs());
+
+        if (!profile.isPresent()) {
+          // Each program run should be associated with a profile, so this should never happen
+          LOG.warn("Will not emit metrics for program {} without profile assigned to it, {}",
+                   programRunId, notification);
+          return Optional.empty();
+        }
+
+        ProfileId profileId = profile.get();
+
+        // emit the metrics information, increment the count for end state program runs
+        MetricsContext metricsContext = getMetricsContextForProfile(programRunId, profileId);
+        metricsContext.increment(getMetricName(meta.getStatus()), 1L);
+
+        // the node number will be null or 0 in local mode, we will not emit node hours for these cases.
+        Integer numNodes = meta.getCluster().getNumNodes();
+        if (numNodes == null || numNodes == 0) {
+          return Optional.empty();
+        }
+
+        // node minutes = (end ts of the cluster - start ts) * nodeNum / 60
+        metricsContext.gauge(Constants.Metrics.Program.PROGRAM_NODE_MINUTES,
```

Comment:
I still have questions about why this is only being done at the end instead of periodically incrementing a counter. This will not give accurate results for longer runs

Reply:
I don't think we should have this anymore, now that there is a counter being incremented.

Reply:
These changes lost when I did the rebase, added in the fix build commit